File: frontend/js/pages/inventory.js
/* Inventory, expiry alerts, batch detail, batch and movement forms. */

import {
    api, badge, daysLabel, debounce, formModal, formatDate, formatDateTime, html, money, mount, number, onAction,
    pageHeader, plural, sortableTable, statTile, table, toast, today,
} from "../core.js";
import { signedQuantity } from "./dashboard.js";

const STATUSES = ["EXPIRED", "CRITICAL", "URGENT", "APPROACHING EXPIRY", "NORMAL"];

async function lookups() {
    const [medicines, locations, suppliers] = await Promise.all([
        api("/medicines", { params: { sort: "name" } }), api("/locations"), api("/suppliers"),
    ]);
    return { medicines, locations: locations.filter(l => l.is_active), suppliers };
}

export async function openBatchForm(medicineId, onSaved) {
    const { medicines, locations, suppliers } = await lookups();
    const fields = [
        { name: "medicine_id", label: "Medicine", type: "select", required: true, value: medicineId, placeholder: "Select…",
          options: medicines.map(m => ({ value: m.id, label: `${m.name} ${m.strength || ""} ${m.dosage_form || ""}` })) },
        { name: "batch_number", label: "Batch number", required: true, maxlength: 100 },
        { name: "quantity", label: "Quantity (units)", type: "number", min: 0, step: 1, required: true },
        { name: "expiry_date", label: "Expiry date", type: "date", required: true },
        { name: "unit_cost", label: "Unit cost", type: "number", min: 0, step: "0.01", help: "Acquisition cost per unit (for valuation)." },
        { name: "location_id", label: "Location", type: "select", value: locations[0]?.id,
          options: locations.map(l => ({ value: l.id, label: l.name })) },
        { name: "supplier_id", label: "Supplier", type: "select", placeholder: "Unknown / none",
          options: suppliers.map(s => ({ value: s.id, label: s.name })) },
        { name: "received_date", label: "Received date", type: "date", value: today() },
    ];
    formModal({
        title: "Register Batch / Opening Stock",
        intro: "Use this for stock not received through a purchase order. Deliveries against an order are received from the order.",
        fields,
        validate: v => (!Number.isInteger(v.quantity) || v.quantity < 0 ? "Quantity must be a whole number of 0 or more." : null),
        onSubmit: async values => {
            ["medicine_id", "location_id", "supplier_id"].forEach(k => { if (values[k] !== null) values[k] = Number(values[k]); });
            await api("/batches", { method: "POST", body: values });
            toast("Batch registered.");
            onSaved?.();
        },
    });
}

const MOVEMENT_HELP = {
    RECEIVED: "Units added to this batch (outside a purchase order).",
    DISPENSED: "Units dispensed from this batch. FEFO is enforced: an earlier-expiring batch must be used first unless overridden.",
    RETURNED: "Units returned to stock (e.g. patient or ward return).",
    DAMAGED: "Units written off as damaged. A reason is required.",
    EXPIRED: "Expired units written off / sent for disposal (only for expired batches).",
    ADJUSTMENT: "Enter the PHYSICALLY COUNTED quantity; the difference is recorded. A reason is required.",
};

export async function openMovementForm(batch, ctx, onSaved) {
    const types = await api("/stock-movements/types");
    const allowed = types.filter(t => t.allowed).map(t => t.movement_type);
    if (!allowed.length) { toast("Your role cannot record stock movements.", "error"); return; }

    const fields = [
        { name: "movement_type", label: "Movement type", type: "select", required: true, value: allowed[0],
          options: allowed.map(t => ({ value: t, label: t })) },
        { name: "quantity", label: "Quantity", type: "number", min: 0, step: 1, required: true,
          help: Object.entries(MOVEMENT_HELP).map(([k, v]) => `${k}: ${v}`).join(" ") },
        { name: "reason", label: "Reason / reference", type: "textarea", full: true },
    ];
    if (ctx.can("stock.fefo_override")) {
        fields.push({ name: "fefo_override_reason", label: "FEFO override reason (only if not dispensing the earliest batch)",
            type: "textarea", full: true });
    }
    formModal({
        title: `Stock movement — ${batch.medicine} batch ${batch.batch_number} (${number(batch.quantity)} in stock)`,
        fields,
        submitLabel: "Record Movement",
        validate: v => (!Number.isInteger(v.quantity) || v.quantity < 0 ? "Quantity must be a whole number." : null),
        onSubmit: async values => {
            const result = await api("/stock-movements", { method: "POST", body: { batch_id: batch.batch_id, ...values } });
            toast(`${result.movement_type} recorded. Batch now holds ${number(result.new_batch_quantity)}.`);
            onSaved?.();
        },
    });
}

/* ---------- Inventory page ---------- */

export async function renderInventory(ctx) {
    const [rows, locations, value] = await Promise.all([
        api("/inventory", { params: { include_empty: false } }), api("/locations"), api("/analytics/valuation"),
    ]);
    if (!ctx.isCurrent()) return;

    mount(ctx.main, html`
        ${pageHeader("Inventory", "Every batch in stock, earliest expiry first", html`
            ${ctx.can("batches.write") ? html`<button type="button" class="refresh-btn primary" data-action="add-batch">+ Register Batch</button>` : ""}
            <button type="button" class="refresh-btn" data-action="refresh">↻ Refresh</button>`)}
        <section class="cards">
            ${statTile("Batches in stock", number(value.total.batches), "blue")}
            ${statTile("Units", number(value.total.units), "green")}
            ${statTile("Stock value", money(value.total.value), "blue",
                value.total.units_without_cost ? `${number(value.total.units_without_cost)} units without cost` : "")}
            ${statTile("Expired value", money(value.expired.value), "red", `${number(value.expired.units)} units`)}
        </section>
        <section class="section">
            <div class="toolbar">
                <input type="search" id="inv-search" placeholder="Search medicine or batch…" aria-label="Search inventory">
                <select id="inv-status" aria-label="Expiry status"><option value="">All expiry statuses</option>
                    ${STATUSES.map(s => html`<option>${s}</option>`)}</select>
                <select id="inv-location" aria-label="Location"><option value="">All locations</option>
                    ${locations.map(l => html`<option value="${l.id}">${l.name}</option>`)}</select>
                <label class="inline-check"><input type="checkbox" id="inv-empty"> Show empty batches</label>
                <span class="toolbar-count" id="inv-count"></span>
            </div>
            <div id="inv-table"></div>
        </section>`);

    const columns = [
        { label: "Medicine", key: "medicine", render: b => html`<a href="#/medicines/${b.medicine_id}"><strong>${b.medicine}</strong></a><br><small>${b.strength} ${b.dosage_form}</small>` },
        { label: "Batch", key: "batch_number", render: b => html`<a href="#/batches/${b.batch_id}">${b.batch_number}</a>` },
        { label: "Qty", key: "quantity", render: b => number(b.quantity), className: "num" },
        { label: "Expiry", key: "expiry_date", render: b => formatDate(b.expiry_date) },
        { label: "Days left", key: "days_until_expiry", render: b => daysLabel(b.days_until_expiry) },
        { label: "Status", key: "status", render: b => badge(b.status), sort: b => STATUSES.indexOf(b.status) },
        { label: "Location", key: "location" },
        { label: "Supplier", key: "supplier", render: b => b.supplier || "—" },
        { label: "Unit cost", key: "unit_cost", render: b => money(b.unit_cost), className: "num", sort: b => Number(b.unit_cost ?? -1) },
        { label: "Value", key: "stock_value", render: b => money(b.stock_value), className: "num", sort: b => Number(b.stock_value ?? -1) },
        { label: "", render: b => html`<button type="button" class="view-btn" data-action="move" data-id="${b.batch_id}">Movement</button>` },
    ];

    let data = rows;
    const search = ctx.main.querySelector("#inv-search");
    const status = ctx.main.querySelector("#inv-status");
    const location = ctx.main.querySelector("#inv-location");
    const empty = ctx.main.querySelector("#inv-empty");
    const draw = () => {
        const term = search.value.trim().toLowerCase();
        const filtered = data.filter(b =>
            (!term || [b.medicine, b.batch_number, b.strength].some(v => (v || "").toLowerCase().includes(term)))
            && (!status.value || b.status === status.value)
            && (!location.value || String(b.location_id) === location.value));
        ctx.main.querySelector("#inv-count").textContent = `${filtered.length} batches`;
        sortableTable(ctx.main.querySelector("#inv-table"), "inventory-table", columns, filtered, { empty: "No batches match." });
    };
    search.addEventListener("input", debounce(draw, 150));
    status.addEventListener("change", draw);
    location.addEventListener("change", draw);
    empty.addEventListener("change", async () => {
        data = await api("/inventory", { params: { include_empty: empty.checked } });
        draw();
    });
    draw();

    onAction(ctx.main, {
        refresh: () => ctx.reload(),
        "add-batch": () => openBatchForm(null, () => ctx.reload()),
        move: el => {
            const batch = data.find(b => b.batch_id === Number(el.dataset.id));
            openMovementForm(batch, ctx, () => ctx.reload());
        },
    });
}

/* ---------- Expiry alerts ---------- */

export async function renderExpiry(ctx) {
    const [alerts, settings] = await Promise.all([api("/expiry-alerts"), api("/settings")]);
    if (!ctx.isCurrent()) return;
    const setting = key => settings.find(s => s.key === key)?.value;
    const s = alerts.summary;

    mount(ctx.main, html`
        ${pageHeader("Expiry Alerts", `Critical ≤ ${setting("expiry.critical_days")} days · Urgent ≤ ${setting("expiry.urgent_days")} days · Approaching ≤ ${setting("expiry.approaching_days")} days`,
            html`<a class="view-btn" href="#/reports">Reports</a>`)}
        <section class="cards">
            ${statTile("Expired", plural(s.EXPIRED.batches, "batch", "batches"), "red", `${number(s.EXPIRED.units)} units · ${money(s.EXPIRED.value)}`)}
            ${statTile("Critical", plural(s.CRITICAL.batches, "batch", "batches"), "red", `${number(s.CRITICAL.units)} units · ${money(s.CRITICAL.value)}`)}
            ${statTile("Urgent", plural(s.URGENT.batches, "batch", "batches"), "orange", `${number(s.URGENT.units)} units · ${money(s.URGENT.value)}`)}
            ${statTile("Approaching", plural(s["APPROACHING EXPIRY"].batches, "batch", "batches"), "blue",
                `${number(s["APPROACHING EXPIRY"].units)} units · ${money(s["APPROACHING EXPIRY"].value)}`)}
        </section>
        <section class="section">
            <div class="toolbar">
                <select id="exp-status" aria-label="Status"><option value="">All alert levels</option>
                    ${STATUSES.slice(0, 4).map(x => html`<option>${x}</option>`)}</select>
            </div>
            <div id="exp-table"></div>
        </section>`);

    const canWriteOff = ctx.can("stock.adjust");
    const columns = [
        { label: "Status", key: "status", render: b => badge(b.status), sort: b => STATUSES.indexOf(b.status) },
        { label: "Medicine", key: "medicine", render: b => html`<a href="#/medicines/${b.medicine_id}"><strong>${b.medicine}</strong></a> <small>${b.strength}</small>` },
        { label: "Batch", key: "batch_number", render: b => html`<a href="#/batches/${b.batch_id}">${b.batch_number}</a>` },
        { label: "Qty", key: "quantity", render: b => number(b.quantity), className: "num" },
        { label: "Expiry", key: "expiry_date", render: b => formatDate(b.expiry_date) },
        { label: "Days", key: "days_until_expiry", render: b => daysLabel(b.days_until_expiry) },
        { label: "Location", key: "location" },
        { label: "Value", key: "stock_value", render: b => money(b.stock_value), className: "num" },
        { label: "", render: b => b.status === "EXPIRED" && canWriteOff
            ? html`<button type="button" class="view-btn danger" data-action="writeoff" data-id="${b.batch_id}">Write off</button>` : "" },
    ];
    const select = ctx.main.querySelector("#exp-status");
    const draw = () => sortableTable(ctx.main.querySelector("#exp-table"), "expiry-table", columns,
        alerts.batches.filter(b => !select.value || b.status === select.value), { empty: "No batches in this category." });
    select.addEventListener("change", draw);
    draw();

    onAction(ctx.main, {
        writeoff: el => {
            const batch = alerts.batches.find(b => b.batch_id === Number(el.dataset.id));
            formModal({
                title: `Write off expired batch ${batch.batch_number}`,
                intro: `${number(batch.quantity)} units of ${batch.medicine} expired on ${formatDate(batch.expiry_date)}. This records an EXPIRED movement and removes the units from stock.`,
                fields: [
                    { name: "quantity", label: "Units to write off", type: "number", min: 1, max: batch.quantity, value: batch.quantity, required: true },
                    { name: "reason", label: "Disposal reference / reason", type: "textarea", required: true, full: true },
                ],
                submitLabel: "Write Off",
                onSubmit: async v => {
                    await api("/stock-movements", { method: "POST", body: { batch_id: batch.batch_id, movement_type: "EXPIRED", ...v } });
                    toast("Expired stock written off.");
                    ctx.reload();
                },
            });
        },
    });
}

/* ---------- Batch detail ---------- */

export async function renderBatch(ctx) {
    const data = await api(`/batches/${encodeURIComponent(ctx.params.id)}`);
    if (!ctx.isCurrent()) return;
    const b = data.batch;

    mount(ctx.main, html`
        ${pageHeader(`Batch ${b.batch_number}`, `${b.medicine} ${b.strength || ""} ${b.dosage_form || ""}`, html`
            <a class="view-btn" href="#/medicines/${b.medicine_id}">← ${b.medicine}</a>
            ${ctx.can("batches.write") ? html`<button type="button" class="refresh-btn" data-action="edit">Edit details</button>` : ""}
            <button type="button" class="refresh-btn primary" data-action="move">Record movement</button>`)}
        <section class="cards">
            <div class="card"><div><span>Quantity</span><strong>${number(b.quantity)}</strong>
                <small class="card-hint">${data.reconciled ? "✓ Matches movement history" : "⚠ Does not match movement history"}</small></div></div>
            <div class="card"><div><span>Expiry</span><strong>${formatDate(b.expiry_date)}</strong><small class="card-hint">${badge(b.status)} ${daysLabel(b.days_until_expiry)}</small></div></div>
            <div class="card"><div><span>Unit cost / value</span><strong>${money(b.unit_cost)}</strong><small class="card-hint">Value ${money(b.stock_value)}</small></div></div>
            <div class="card"><div><span>Location · Supplier</span><strong>${b.location}</strong><small class="card-hint">${b.supplier || "Supplier unknown"} · received ${formatDate(b.received_date)}</small></div></div>
        </section>
        <section class="section">
            <div class="section-header"><div><h3>Batch history</h3><p>Every movement with the running balance</p></div></div>
            ${table("batch-history", [
                { label: "Date", render: m => formatDateTime(m.movement_date) },
                { label: "Type", render: m => badge(m.movement_type) },
                { label: "Change", render: m => signedQuantity(m), className: "num" },
                { label: "Balance", render: m => number(m.balance_after), className: "num" },
                { label: "Reason", render: m => m.reason || "—" },
                { label: "By", render: m => m.user_name || "—" },
            ], data.history, { empty: "No movements recorded." })}
        </section>
        ${data.receipts.length ? html`<section class="section">
            <div class="section-header"><div><h3>Purchase receipts</h3></div></div>
            ${table("batch-receipts", [
                { label: "Received", render: r => formatDateTime(r.received_date) },
                { label: "Order", render: r => html`<a href="#/purchasing/${r.purchase_order_id}">${r.order_number}</a>` },
                { label: "Qty", render: r => number(r.quantity_received) },
                { label: "Unit cost", render: r => money(r.unit_cost) },
                { label: "Received by", render: r => r.received_by || "—" },
            ], data.receipts)}</section>` : ""}`);

    onAction(ctx.main, {
        move: () => openMovementForm(b, ctx, () => ctx.reload()),
        edit: async () => {
            const suppliers = await api("/suppliers");
            formModal({
                title: `Edit batch ${b.batch_number}`,
                intro: "Correct batch details. Quantity changes only through stock movements.",
                fields: [
                    { name: "batch_number", label: "Batch number", required: true, value: b.batch_number, maxlength: 100 },
                    { name: "expiry_date", label: "Expiry date", type: "date", required: true, value: b.expiry_date },
                    { name: "unit_cost", label: "Unit cost", type: "number", min: 0, step: "0.01", value: b.unit_cost },
                    { name: "supplier_id", label: "Supplier", type: "select", placeholder: "Unknown / none", value: b.supplier_id,
                      options: suppliers.map(s => ({ value: s.id, label: s.name })) },
                    { name: "received_date", label: "Received date", type: "date", value: b.received_date },
                    { name: "reason", label: "Reason for change", type: "textarea", required: true, full: true },
                ],
                onSubmit: async v => {
                    if (v.supplier_id !== null) v.supplier_id = Number(v.supplier_id);
                    await api(`/batches/${b.batch_id}`, { method: "PUT", body: v });
                    toast("Batch updated.");
                    ctx.reload();
                },
            });
        },
    });
}

File: backend/routers/inventory.py
# ============================================================
# INVENTORY, BATCHES, STOCK ALERTS, FEFO AND DISPENSING
# ============================================================
# Original endpoints kept: GET /inventory, POST /batches, GET /stock-alerts.
# Their original fields are unchanged; new fields are additions.

from datetime import date
from typing import Literal

import psycopg
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from .. import audit
from ..database import get_db
from ..schemas import Code100, LongText, blank_to_none
from ..security import CurrentUser, require
from ..services import expiry, inventory, stock

router = APIRouter(tags=["Inventory"])

ExpiryStatus = Literal["EXPIRED", "CRITICAL", "URGENT", "APPROACHING EXPIRY", "NORMAL"]


class BatchCreate(BaseModel):
    medicine_id: int
    batch_number: Code100
    quantity: int = Field(ge=0, le=100_000_000)
    expiry_date: date
    unit_cost: float | None = Field(default=None, ge=0, le=10_000_000)
    location_id: int | None = None
    supplier_id: int | None = None
    received_date: date | None = None


class BatchUpdate(BaseModel):
    batch_number: Code100
    expiry_date: date
    unit_cost: float | None = Field(default=None, ge=0, le=10_000_000)
    supplier_id: int | None = None
    received_date: date | None = None
    reason: LongText = Field(min_length=3)


class DispenseRequest(BaseModel):
    medicine_id: int
    quantity: int = Field(gt=0, le=10_000_000)
    location_id: int | None = None
    reason: LongText | None = None


def default_location_id(conn) -> int:
    row = conn.execute(
        "SELECT id FROM locations WHERE is_active ORDER BY id LIMIT 1"
    ).fetchone()
    if row is None:
        raise HTTPException(status_code=400, detail="No active stock location exists")
    return row["id"]


def check_location(conn, location_id: int) -> None:
    row = conn.execute("SELECT is_active FROM locations WHERE id = %s", (location_id,)).fetchone()
    if row is None:
        raise HTTPException(status_code=404, detail="Location not found")
    if not row["is_active"]:
        raise HTTPException(status_code=400, detail="Location is inactive")


def check_supplier(conn, supplier_id: int) -> None:
    if conn.execute("SELECT 1 FROM suppliers WHERE id = %s", (supplier_id,)).fetchone() is None:
        raise HTTPException(status_code=404, detail="Supplier not found")


# ------------------------------------------------------------
# Inventory
# ------------------------------------------------------------

@router.get("/inventory")
def get_inventory(
    search: str | None = Query(default=None, max_length=100),
    status: ExpiryStatus | None = None,
    medicine_id: int | None = None,
    location_id: int | None = None,
    supplier_id: int | None = None,
    include_empty: bool = True,
    user: CurrentUser = Depends(require("inventory.read")),
    conn: psycopg.Connection = Depends(get_db),
):
    rows = inventory.batches(
        conn, search=search, status=status, medicine_id=medicine_id, location_id=location_id,
        supplier_id=supplier_id, include_empty=include_empty,
    )
    # Original fields first (expiry_date as a string, as before), then additions.
    return [{**row, "expiry_date": str(row["expiry_date"])} for row in rows]


@router.get("/expiry-alerts")
def expiry_alerts(user: CurrentUser = Depends(require("inventory.read")),
                  conn: psycopg.Connection = Depends(get_db)):
    """Batches with stock that are expired or inside a warning threshold."""
    rows = [r for r in inventory.batches(conn, include_empty=False) if r["status"] != expiry.NORMAL]
    summary = {}
    for status_name in expiry.STATUS_ORDER[:-1]:
        group = [r for r in rows if r["status"] == status_name]
        summary[status_name] = {
            "batches": len(group),
            "units": sum(r["quantity"] for r in group),
            "value": round(sum(float(r["stock_value"] or 0) for r in group), 2),
        }
    return {"summary": summary, "batches": rows}


# ------------------------------------------------------------
# Batches
# ------------------------------------------------------------

@router.post("/batches")
def create_batch(batch: BatchCreate, user: CurrentUser = Depends(require("batches.write")),
                 conn: psycopg.Connection = Depends(get_db)):
    """Register a batch directly (opening stock, donation, stock outside a
    purchase order). The initial quantity is recorded as a RECEIVED movement
    so the stock ledger stays reconciled."""
    if conn.execute("SELECT 1 FROM medicines WHERE id = %s", (batch.medicine_id,)).fetchone() is None:
        raise HTTPException(status_code=404, detail="Medicine not found")

    location_id = batch.location_id or default_location_id(conn)
    check_location(conn, location_id)
    if batch.supplier_id is not None:
        check_supplier(conn, batch.supplier_id)

    exists = conn.execute(
        "SELECT id FROM batches WHERE medicine_id = %s AND batch_number = %s AND location_id = %s",
        (batch.medicine_id, batch.batch_number, location_id),
    ).fetchone()
    if exists:
        raise HTTPException(
            status_code=409,
            detail=f"Batch {batch.batch_number} already exists for this medicine at this location "
                   f"(batch id {exists['id']}). Record a RECEIVED movement against it instead.",
        )

    row = conn.execute(
        """
        INSERT INTO batches (medicine_id, batch_number, quantity, expiry_date, location_id,
                             unit_cost, supplier_id, received_date)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
        RETURNING id, medicine_id, batch_number, quantity, expiry_date, location_id, unit_cost,
                  supplier_id, received_date
        """,
        (batch.medicine_id, batch.batch_number, batch.quantity, batch.expiry_date, location_id,
         batch.unit_cost, batch.supplier_id, batch.received_date or date.today()),
    ).fetchone()

    if batch.quantity > 0:
        stock.insert_movement(conn, row["id"], "RECEIVED", batch.quantity,
                              "Batch registered (stock entered directly)", user)

    audit.record(conn, user, "CREATE", "batch", row["id"], None, dict(row))
    conn.commit()
    return {**row, "expiry_date": str(row["expiry_date"])}


@router.get("/batches/{batch_id}")
def batch_detail(batch_id: int, user: CurrentUser = Depends(require("inventory.read")),
                 conn: psycopg.Connection = Depends(get_db)):
    rows = inventory.batches(conn, batch_id=batch_id)
    if not rows:
        raise HTTPException(status_code=404, detail="Batch not found")

    history = conn.execute(
        """
        SELECT sm.id, sm.movement_type, sm.quantity, sm.movement_date, sm.reason,
               users.full_name AS user_name
        FROM stock_movements sm
        LEFT JOIN users ON users.id = sm.user_id
        WHERE sm.batch_id = %s
        ORDER BY sm.movement_date, sm.id
        """,
        (batch_id,),
    ).fetchall()

    balance = 0
    for movement in history:
        sign = 1 if movement["movement_type"] in ("RECEIVED", "RETURNED", "ADJUSTMENT") else -1
        balance += sign * movement["quantity"]
        movement["balance_after"] = balance

    receipts = conn.execute(
        """
        SELECT pr.id, pr.received_date, pr.quantity_received, pr.received_by,
               po.id AS purchase_order_id, po.order_number, poi.unit_cost
        FROM purchase_receipts pr
        JOIN purchase_order_items poi ON poi.id = pr.purchase_order_item_id
        JOIN purchase_orders po ON po.id = poi.purchase_order_id
        WHERE pr.batch_id = %s
        ORDER BY pr.received_date
        """,
        (batch_id,),
    ).fetchall()

    return {
        "batch": rows[0],
        "history": history,
        "ledger_balance": balance,
        "reconciled": balance == rows[0]["quantity"],
        "receipts": receipts,
    }


@router.put("/batches/{batch_id}")
def update_batch(batch_id: int, body: BatchUpdate, user: CurrentUser = Depends(require("batches.write")),
                 conn: psycopg.Connection = Depends(get_db)):
    """Correct batch details. Quantity is never edited here: it changes only
    through stock movements, so the ledger stays complete."""
    old = conn.execute(
        """
        SELECT id, medicine_id, location_id, batch_number, expiry_date, unit_cost, supplier_id, received_date
        FROM batches WHERE id = %s FOR UPDATE
        """,
        (batch_id,),
    ).fetchone()
    if old is None:
        raise HTTPException(status_code=404, detail="Batch not found")
    if body.supplier_id is not None:
        check_supplier(conn, body.supplier_id)

    clash = conn.execute(
        """
        SELECT id FROM batches
        WHERE medicine_id = %s AND batch_number = %s AND location_id = %s AND id <> %s
        """,
        (old["medicine_id"], body.batch_number, old["location_id"], batch_id),
    ).fetchone()
    if clash:
        raise HTTPException(status_code=409, detail="Another batch with this number exists at this location")

    row = conn.execute(
        """
        UPDATE batches SET batch_number = %s, expiry_date = %s, unit_cost = %s, supplier_id = %s,
               received_date = %s
        WHERE id = %s
        RETURNING id, batch_number, expiry_date, unit_cost, supplier_id, received_date
        """,
        (body.batch_number, body.expiry_date, body.unit_cost, body.supplier_id, body.received_date, batch_id),
    ).fetchone()

    fields = ["batch_number", "expiry_date", "unit_cost", "supplier_id", "received_date"]
    before, after = audit.changed_fields({k: old[k] for k in fields}, {k: row[k] for k in fields})
    if after:
        after["reason"] = body.reason
        audit.record(conn, user, "UPDATE", "batch", batch_id, before, after)
    conn.commit()
    return row


# ------------------------------------------------------------
# Stock levels (low stock)
# ------------------------------------------------------------

@router.get("/stock-alerts")
def get_stock_alerts(location_id: int | None = None,
                     user: CurrentUser = Depends(require("inventory.read")),
                     conn: psycopg.Connection = Depends(get_db)):
    """Per-medicine stock levels.

    current_stock is USABLE stock (expired batches excluded, since they
    cannot be dispensed); expired_stock and total_stock are listed
    separately. status is OUT OF STOCK / LOW STOCK / NORMAL.
    """
    rows = inventory.medicine_stock(conn, location_id=location_id)
    return [
        {
            "medicine_id": row["medicine_id"],
            "medicine": row["medicine"],
            "strength": row["strength"],
            "dosage_form": row["dosage_form"],
            "current_stock": row["usable_stock"],
            "reorder_level": row["reorder_level"],
            "status": row["stock_status"],
            "expired_stock": row["expired_stock"],
            "total_stock": row["total_stock"],
            "batches_in_stock": row["batches_in_stock"],
            "next_expiry": row["next_expiry"],
            "stock_value": row["stock_value"],
        }
        for row in rows
    ]


# ------------------------------------------------------------
# FEFO and dispensing
# ------------------------------------------------------------

@router.get("/fefo/{medicine_id}")
def fefo_preview(medicine_id: int, quantity: int = Query(default=1, gt=0, le=10_000_000),
                 location_id: int | None = None,
                 user: CurrentUser = Depends(require("inventory.read")),
                 conn: psycopg.Connection = Depends(get_db)):
    """Which batches FEFO would use to supply `quantity` units (no changes made)."""
    if conn.execute("SELECT 1 FROM medicines WHERE id = %s", (medicine_id,)).fetchone() is None:
        raise HTTPException(status_code=404, detail="Medicine not found")
    return stock.fefo_plan(conn, medicine_id, quantity, location_id)


@router.post("/dispense")
def dispense(body: DispenseRequest, user: CurrentUser = Depends(require("stock.dispense")),
             conn: psycopg.Connection = Depends(get_db)):
    """Dispense a quantity of a medicine using FEFO across batches."""
    if body.location_id is not None:
        check_location(conn, body.location_id)
    result = stock.dispense_fefo(conn, user, body.medicine_id, body.quantity, body.location_id,
                                 blank_to_none(body.reason))
    conn.commit()
    return result


@router.get("/stock-reconciliation")
def stock_reconciliation(user: CurrentUser = Depends(require("inventory.read")),
                         conn: psycopg.Connection = Depends(get_db)):
    """Batches whose quantity does not match their movement history. Empty
    means the ledger is fully reconciled."""
    mismatches = stock.reconciliation(conn)
    return {"reconciled": not mismatches, "mismatches": mismatches}
